File: Sources/PalmierPro/Models/GradeCurve.swift
import CoreImage
import Foundation

struct CurvePoint: Codable, Sendable, Equatable {
    var x: Double
    var y: Double
}

/// Master (Rec.709 luma) + per-channel R/G/B tone curves, compiled to a `CIColorCube`.
struct GradeCurve: Codable, Sendable, Equatable {
    var master: [CurvePoint] = []
    var red: [CurvePoint] = []
    var green: [CurvePoint] = []
    var blue: [CurvePoint] = []

    static let identityPoints = [CurvePoint(x: 0, y: 0), CurvePoint(x: 1, y: 1)]

    var isIdentity: Bool {
        [master, red, green, blue].allSatisfy { $0.isEmpty || $0 == Self.identityPoints }
    }

    /// Piecewise-linear interpolation, clamped flat outside the point range.
    static func eval(_ pts: [CurvePoint], _ x: Double) -> Double {
        let p = (pts.isEmpty ? identityPoints : pts).sorted { $0.x < $1.x }
        if x <= p[0].x { return p[0].y }
        if x >= p[p.count - 1].x { return p[p.count - 1].y }
        for i in 1..<p.count where x <= p[i].x {
            let a = p[i - 1], b = p[i]
            let t = (b.x - a.x) == 0 ? 0 : (x - a.x) / (b.x - a.x)
            return a.y + (b.y - a.y) * t
        }
        return x
    }

    func encoded() -> String? {
        (try? JSONEncoder().encode(self)).flatMap { String(data: $0, encoding: .utf8) }
    }
}

/// Failable JSON init kept in an extension so the memberwise initializer survives.
extension GradeCurve {
    init?(json: String) {
        guard let data = json.data(using: .utf8),
              let decoded = try? JSONDecoder().decode(GradeCurve.self, from: data) else { return nil }
        self = decoded
    }
}

File: windows/src/PalmierPro.Core/Models/GradeCurve.cs
using PalmierPro.Core.Serialization;

namespace PalmierPro.Core.Models;

public record struct CurvePoint(double X, double Y);

/// <summary>Master (Rec.709 luma) + per-channel R/G/B tone curves.</summary>
public sealed class GradeCurve
{
    public List<CurvePoint> Master { get; set; } = [];
    public List<CurvePoint> Red { get; set; } = [];
    public List<CurvePoint> Green { get; set; } = [];
    public List<CurvePoint> Blue { get; set; } = [];

    public static readonly IReadOnlyList<CurvePoint> IdentityPoints = [new(0, 0), new(1, 1)];

    public bool IsIdentity =>
        new[] { Master, Red, Green, Blue }
            .All(points => points.Count == 0 || points.SequenceEqual(IdentityPoints));

    /// <summary>Piecewise-linear interpolation, clamped flat outside the point range.</summary>
    public static double Eval(IReadOnlyList<CurvePoint> pts, double x)
    {
        var p = (pts.Count == 0 ? IdentityPoints : pts).OrderBy(pt => pt.X).ToList();
        if (x <= p[0].X) return p[0].Y;
        if (x >= p[^1].X) return p[^1].Y;
        for (var i = 1; i < p.Count; i++)
        {
            if (x > p[i].X) continue;
            var a = p[i - 1];
            var b = p[i];
            var t = b.X - a.X == 0 ? 0 : (x - a.X) / (b.X - a.X);
            return a.Y + (b.Y - a.Y) * t;
        }
        return x;
    }

    public string? Encoded()
    {
        try
        {
            return PalmierJson.EncodeToString(this);
        }
        catch
        {
            return null;
        }
    }

    public static GradeCurve? FromJson(string json)
    {
        try
        {
            return PalmierJson.Decode<GradeCurve>(json);
        }
        catch
        {
            return null;
        }
    }
}
